### ISSUE TITLE
Filtering in nested connection fields breaks optimizer

### ISSUE THREAD

**Issue by rsomani95** [MrThearMan/graphene-django-query-optimizer#62]
### I have read the docs thoroughly before making this bug report.

- [X] Yes

### I have read through other open issues, and my issue is not a duplicate.

- [X] Yes

### What version of the library you are using?

0.1.11

### Which python version are you using?

3.12

### What operating system are you on?

Mac

### Description

For nested fields, the filtering arguments aren't being registered in the GQL schema. My setup is as follows (partly copied from the [test models](https://github.com/MrThearMan/graphene-django-query-optimizer/blob/f13e9fc037cdb4999f449fdbcb5cfbc9639f02c1/tests/example/types.py#L268-L309):

```python
class HousingCompanyFilterSet(FilterSet):
    order_by = OrderingFilter(fields=["name"])

    class Meta:
        model = HousingCompany
        fields = {
            "name": ["iexact", "icontains"]
        }

class HousingCompanyNode(DjangoObjectType):
    real_estates = DjangoConnectionField(RealEstateNode)

    def resolve_real_estates(self, info, *args, **kwargs):
        return self.real_estates.all()

    class Meta:
        model = HousingCompany
        interfaces = (relay.Node,)
        filterset_class = HousingCompanyFilterSet


class PropertyManagerFilterSet(FilterSet):
    class Meta:
        model = PropertyManager
        fields = ["name", "email"]
```

Then, I have a `PropertyManagerNode` setup as follows:

```python
class PropertyManagerNode(DjangoObjectType):
    housing_companies = DjangoConnectionField(HousingCompanyNode)
```

**GOAL:** I'd like to be able to query for all the `PropertyManager`s who have `HousingCompany`s that match a certain name. Here's how I'm trying to achieve that:

```graphql
query {
  pagedPropertyManagers {
    edges {
      node {
        id
        housingCompanies(
          name: "sons"  # Unable to do this
        ) {
          edges {
            node {
              name
            }
          }
        }
      }
    }
  }
}
```

However, I'm unable to filter by `name` on `housingCompanies` (the nested item) when doing this query. The only registered filters are the relay pagination filters.

Out of curiousity, I changed the definition to explicitly be a `DjangoFilterConnectionField`, like so:

```python
class PropertyManagerNode(DjangoObjectType):
    housing_companies = DjangoFilterConnectionField(HousingCompanyNode)
```

But this causes N+1 queries because the [`ConnectionFieldCachingMixin`](https://github.com/MrThearMan/graphene-django-query-optimizer/blob/f13e9fc037cdb4999f449fdbcb5cfbc9639f02c1/query_optimizer/fields.py#L46) is not being applied.

---

Looking at the `query_optimizer.fields` module, it's clear that the `DjangoConnectionField` is [implicitly applied](https://github.com/MrThearMan/graphene-django-query-optimizer/blob/f13e9fc037cdb4999f449fdbcb5cfbc9639f02c1/query_optimizer/fields.py#L144-L166) for `ManyToManyField`, `ManyToManyRel` and `ManyToOneRel`, and as I understand it (per [this SO](https://stackoverflow.com/questions/15300422/difference-between-manytoonerel-and-foreignkey) post and other reading), `ManyToOneRel` is the "other" side of the `ForeignKey` relationship used internally by Django. If I understand correctly, that's the relationship between `PropertyManager` and `HousingCompany`. So perhaps there's a bug somewhere that fails to trigger the registration of this relationship?

Worth mentioning that the filters are available if I'm querying for `pagedHousingCompanies` at the top level. This issue only occurs at the nested level.

**Comment by MrThearMan** [comment]
Was able to recreate this locally.

One note at the beginning about `DjangoConnectionField` vs. `DjangoFilterConnectionField`. The former only includes pagination, while the latter also includes filters defined by the object types `Meta.filterset_class` or `Meta.filter_fields`. However, using the latter when neither `Meta.filterset_class` or `Meta.filter_fields` exist will raise an error. If you don't explicitly declate the field, `query_optimizer.fields.convert_to_many_field_to_list_or_connection` will determine the correct field to use based on if filters are included or not.

Now, it seems like the issue happens due to the connection field resolver losing it's result cache, and when it goes though the nested connection resolver, it makes a [count query here](https://github.com/graphql-python/graphene-django/blob/54372b41d501229abdf21d8109a548fe60772b1a/graphene_django/fields.py#L159). This results in N+1 count queries for the related entities:

```shell
1) ------------------------------------------------------------------------
SELECT COUNT(*) AS "__count"
FROM "example_propertymanager"
2) ------------------------------------------------------------------------
SELECT "example_propertymanager"."id"
FROM "example_propertymanager"
ORDER BY "example_propertymanager"."name" ASC
LIMIT 10
3) ------------------------------------------------------------------------
SELECT "example_housingcompany"."id",
       "example_housingcompany"."name",
       "example_housingcompany"."property_manager_id"
FROM "example_housingcompany"
WHERE "example_housingcompany"."property_manager_id" IN (...)
ORDER BY "example_housingcompany"."name" ASC
4) ------------------------------------------------------------------------
SELECT COUNT(*) AS "__count"
FROM "example_housingcompany"
WHERE ("example_housingcompany"."property_manager_id" = ...
       AND "example_housingcompany"."name" LIKE sons ESCAPE '\')
5) ------------------------------------------------------------------------
SELECT COUNT(*) AS "__count"
FROM "example_housingcompany"
WHERE ("example_housingcompany"."property_manager_id" = ...
       AND "example_housingcompany"."name" LIKE sons ESCAPE '\')
6) ------------------------------------------------------------------------
SELECT COUNT(*) AS "__count"
FROM "example_housingcompany"
WHERE ("example_housingcompany"."property_manager_id" = ...
       AND "example_housingcompany"."name" LIKE sons ESCAPE '\')
7) ------------------------------------------------------------------------

 ect.
```

I'll need to track down where the queryset result cache gets cleared, so that the count query can use the result cache.

**Comment by rsomani95** [comment]
Ahhh, great catch!

> One note at the beginning about DjangoConnectionField vs. DjangoFilterConnectionField. The former only includes pagination, while the latter also includes filters defined by the object types Meta.filterset_class or Meta.filter_fields. However, using the latter when neither Meta.filterset_class or Meta.filter_fields exist will raise an error. If you don't explicitly declate the field, query_optimizer.fields.convert_to_many_field_to_list_or_connection will determine the correct field to use based on if filters are included or not.

Yup, I ran into this but didn't bring it up as I wanted to stay focused on the issue. If you think it's worth adding to the docs, I'll be happy to put in a PR.

**Comment by MrThearMan** [comment]
Update on this. It seems like the issue is more than just extra count queries being made. My previous observation was just because I was not receiving any nesed fields based on my filters.

However, it seems like this is going to require some re-designing of the optimizer to support seemlessly. I started by crating a connection field from scratch, since it was getting hard to jerry-rig the existing field. I made mine work with both filtered and non-filtered object types, so there shoud no longer be a need to change the field based on whether filters are added or not. It doesn't support all the features the old one did (for simplicity), but IMO I only removed options that shouldn't be in common use. Unless there is demand for any specific one, I'd leave paching in the old behavior for library users.

Next, I'll need to think about a way to do all the filterset filtering early so that the filters are applied to the prefetches correctly. This should fix the issues, but I'm not sure how I'll be able to find the correct filterset for each child relation yet.

**Comment by rsomani95** [comment]
This all sounds great. Exciting stuff, thanks for the update!

**Comment by MrThearMan** [comment]
Ok, merged changes to [main](https://github.com/MrThearMan/graphene-django-query-optimizer/tree/355b50783cc239c76ca1dbeb11ab515d87197c2c) which should fix this issue. I refactored a bunch of stuff at the same time to make this work, but in the end I'm much more happy with how things are structured now. I'll still try to get the prefetch limiting from #61 working before I make a new release, which will be 0.2.0. Should be done this week.

**Comment by MrThearMan** [comment]
Released in [0.2.0](https://github.com/MrThearMan/graphene-django-query-optimizer/releases/tag/v0.2.0)
